docs/develop/python/message-passing.mdx

```diff
@@ -221,132 +237,255 @@ async def main():
+  This is the error you'll see if the Update was rejected by the validator, or the update was failed after having been accepted.
 
-- **Decorator Usage:** Apply `@workflow.update` to the method intended to handle updates.
-- **Overriding:** If a method with this decorator is overridden, the overriding method should also be decorated with `@workflow.update`.
-- **Validator Method:** Optionally, you can define a validator method for the update handler. This validator is specified using `@update_handler_method_name.validator` and is invoked before the update handler.
-- **Method Parameters:** Update handlers should only use positional parameters. For non-dynamic methods, it's recommended to use a single parameter that is an object or data class, which allows for future expansion of fields.
-- **Return Values:** The update handler can return a serializable value. This value is sent back to the caller of the update.
+- [`temporalio.service.RPCError`](https://python.temporal.io/temporalio.service.RPCError.html) "workflow execution already completed"`
+
+  This will happen if the Workflow finished while the Update handler execution was in progress, for example because
+  - The Workflow was canceled or was deliberately failed (the Workflow author raised [`ApplicationError`](https://python.temporal.io/temporalio.exceptions.ApplicationError.html) outside an Update handler)
+  - The Workflow completed normally or continued-as-new and the Workflow author did not [wait for handlers to be finished](/encyclopedia/workflow-message-passing#finishing-message-handlers).
+
+If the Workflow handle references a Workflow that doesn't exist then [`execute_update`](https://python.temporal.io/temporalio.client.WorkflowHandle.html#execute_update) and [`start_update`](https://python.temporal.io/temporalio.client.WorkflowHandle.html#start_update) will both raise `temporalio.service.RPCError "sql: no rows in result set"`.
+
+
+## Message handler patterns {#message-handler-patterns}
+
+In addition to the topics below, see [Inject work into the main Workflow](/encyclopedia/workflow-message-passing#injecting-work-into-main-workflow) and [Ensuring your messages are processed exactly once](/encyclopedia/workflow-message-passing#exactly-once-message-processing).
+
+### Async handlers {#async-handlers}
+In the example at the top of the page, all Signal and Update handlers were plain `def` methods.
+But handlers can also be `async def`.
+This allows them to use `await` to wait for Activities, Child Workflows, [`asyncio.sleep`](https://docs.python.org/3/library/asyncio-task.html#asyncio.sleep) timers, [`workflow.wait_condition`](https://python.temporal.io/temporalio.workflow.html#wait_condition) conditions, etc, thus opening up many powerful possibilities.
+However, this means that handler executions, and your main Workflow method, are all running concurrently, with switching occurring between them at `await` calls.
+(I.e. they interleave, but there is no parallelism.)
```

Comment:
I would try to limit the use of parenthesis here, especially since a capitalized "I.e." looks a bit strange.

Reply:
Thanks, that got done.